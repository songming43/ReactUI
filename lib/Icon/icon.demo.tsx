import React from "react";
import Demo from "../../demo";
import IconExample from "./icon.example";
// tslint:disable-next-line: no-var-requires
const code = require("!!raw-loader!./icon.example.tsx");

const IconDemo = () => {
  return (
    <Demo code={code.default}>
      <IconExample />
    </Demo>
  );
};

export default IconDemo;
