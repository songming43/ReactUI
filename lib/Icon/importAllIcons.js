let importAll = (requireContext) => requireContext.keys().forEach(requireContext)
try {
  importAll(require.context('../../assert/icons/', true, /\.svg$/))
} catch (error) {
}
