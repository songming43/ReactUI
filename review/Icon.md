## Icon

### 设计 API
```html
<Icon name="xxx">
```

```ts
interface IconProps {
  name: string
}

/** 
 * <> 代表这个类型接受一个参数, 不传代表无限制
 */
const Icon: React.FunctionComponent<IconProps> = (props) => {}
```


### icon 使用 svg 格式
- webpack 使用 `svg-sprite-loader`
- `tsconfig.json` 的 inclued 包含 那些 ts 识别
- `lib/**/*` 代表 lib 文件夹下的所有文件

```text
icon 使用 svg 标签
```

```ts
// * 使用 import from时, 需要让 svg 默认导出
// 让svg 有一个单独的导出
declare module '*.svg' {
  const content: any;
  export default content;
}

```


### 动态加载 很多 icon
```js
let importAll = (requireContext) => requireContext.keys().forEach(requireContext)
try {
  importAll(require.context('../../assets/icons/', true, /\.svg$/))
} catch (error) {
}

```

### tree-shaking
- 使用 importAll 一次性引入, 无法使用 tree-shaking技术
