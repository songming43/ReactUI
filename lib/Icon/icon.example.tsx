import React from "react";
import Icon from "./icon";

const IconExample: React.FunctionComponent = () => {
  const fn: React.MouseEventHandler = (e) => {
    console.log(e);
    console.log(e.target);
  }

  return (
    <div>
      <Icon name="message" color="blue" />
      <Icon name="view" />
      <Icon name="view_off" className="icon" />
      <Icon name="edit" className="icon" color="red" />
      <Icon name="close" className="icon" color="red" size="20px" />
      <Icon name="upload" onClick={fn} />
    </div>
  );
};

export default IconExample;
