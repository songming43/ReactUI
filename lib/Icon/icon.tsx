import React from "react";
import "./importAllIcons";
import "./icon.scss";
import classes from "../helpers/classes";

interface IconProps extends React.SVGAttributes<SVGElement> {
  name: string;
}

const Icon: React.FunctionComponent<IconProps> = ({
  className,
  name,
  ...restProps
}) => {
  return (
    <svg className={classes("yui-icon", className)} {...restProps}>
      <use xlinkHref={`#${name}`}></use>
    </svg>
  );
};

export default Icon;
