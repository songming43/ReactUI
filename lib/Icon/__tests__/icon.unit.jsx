import * as renderer from 'react-test-renderer'
import * as React from 'react'
import Icon from '../icon'
import { mount } from  'enzyme'

describe('测试Icon', () => {
  it('render success', () => {
    const json = renderer.create(<Icon name="qq" />).toJSON()
    expect(json).toMatchSnapshot()
  })
  it('onClick', () => {
    let n = 1
    const fn = () => {
      n = 2
    }
    const component = mount(<Icon name="qq" onClick={fn} />)
    component.find('svg').simulate('click')
    expect(n).toEqual(2)
  })
  it('onClick jest.fn()', () => {
    const fn = jest.fn()
    const component = mount(<Icon name="qq" onClick={fn} />)
    component.find('svg').simulate('click')
    expect(fn).toBeCalled()
  })
});
