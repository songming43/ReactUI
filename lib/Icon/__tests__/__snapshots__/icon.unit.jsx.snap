// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试Icon render success 1`] = `
<svg
  className="yui-icon"
>
  <use
    xlinkHref="#qq"
  />
</svg>
`;
